include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+
+        return regenerateEntryAt(i);
+      }
+
+      return true;
+    }
+
+    bool regenerateEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        entry.shrinkable = genFunc()(entry.random, size);
+        const auto cmd = entry.shrinkable.value();
+        return true;
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+      } catch (const GenerationFailure &failure) {
+        // Just return false below
+      }
+
+      return false;
+    }
+  };
+
+  struct ParallelCommandSequence {
+    ParallelCommandSequence(CommandSequence prefix,
+                            CommandSequence left,
+                            CommandSequence right)
+        : prefix(prefix)
+        , left(left)
+        , right(right) {}
+
+    CommandSequence prefix;
+    CommandSequence left;
+    CommandSequence right;
+  };
+
+  Shrinkable<ParallelCommands<Cmd>> generateCommands(const Random &random,
+                                                     int size) const {
+    return shrinkable::map(generateParallelSequence(random, size),
+                           [](const ParallelCommandSequence &sequence) {
+                             ParallelCommands<Cmd> cmds;
+                             copyCommandEntries(sequence.prefix, cmds.prefix);
+                             copyCommandEntries(sequence.left, cmds.left);
+                             copyCommandEntries(sequence.right, cmds.right);
+                             return cmds;
+                           });
+  }
+
+  template <typename T>
+  static void copyCommandEntries(const CommandSequence &source,
+                                 std::vector<T> &dest) {
+    const auto &entries = source.entries;
+    dest.reserve(entries.size());
+    std::transform(
+        begin(entries),
+        end(entries),
+        std::back_inserter(dest),
+        [](const CommandEntry &entry) { return entry.shrinkable.value(); });
+  }
+
+  Shrinkable<ParallelCommandSequence>
+  generateParallelSequence(const Random &random, int size) const {
+    Random r(random);
+    std::size_t count = (r.split().next() % (size + 1)) + 1;
+    return shrinkable::shrinkRecur(generateInitialParallel(random, size),
+                                   &shrinkSequence);
+  }
+
+  ParallelCommandSequence generateInitialParallel(const Random &random,
+                                                  int size) const {
+    auto r = random;
+    int prefixSz, leftSz, rightSz;
+    std::tie(prefixSz, leftSz, rightSz) = parallelCommandDistribution(size);
+
+    std::size_t prefixCount = (r.split().next() % (prefixSz + 1));
+    std::size_t leftCount = (r.split().next() % (leftSz + 1));
+    std::size_t rightCount = (r.split().next() % (rightSz + 1));
+
```

Comment:
To get a better mental model of why we need to do splitting, let's say we want to generate three vectors of random numbers from a single `Random` without splitting:

```
a = [5, 2, 10]
b = [2, 5, 1, 20]
c = [7, 5, 3]
```

The `Random` returned `5, 2, 10, 2, 5, 1, 20, 7, 5, 3` in sequence. Let's say we apply some shrinking to `b` so suddenly, it doesn't need to pick as many random numbers any more. If the same sequence is returned, we would get this:

```
a = [5, 2, 10]
b = [2, 5, 1]
c = [20, 7, 5]
```

Oops, modifying `b` also made `c` change because they use the same `Random` to generate their values.

What `.split()` does is give us two random generators that are completely independent:

```
Random r1;
r2 = r1.split();
```

`r1` and `r2` is now independent. `split` modifies `r1` and turns it into a `Random` that is independent from the returned one.

However, in the case above, there is no need to split each since we know that we always pick exactly three values. We still need to split it once since we cannot call `split` after we've called `next` and we need a generator for below.